### PR TITLE
Improve turn channel classification: detect slip lanes

### PR DESCRIPTION
**Pull request by merkispavel**
# Issue

This PR changes logic of the ramp-vs-turnChannel classification, i.e
```python
# Before
if IsTurnChannel(links) and length(links) < threshold:
  turn_channel = true

# After
if IsTurnChannel(links) and (length(links) < threshold or IsSlipLane(links)):
  turn_channel = true
```

So a fallback check is added for the cases when length criteria is not satisfied. See the comments in the `linkclassification.cc` for `IsSlipLane` implementation details (https://github.com/valhalla/valhalla/pull/3196/files#r666282023)

## Screenshots
Legend:
- orange lines = turnChannels
- blue lines = ramps

| Before | After |
| ------------- | ------------- |
|   Screenshot 2021-07-08 at 17 36 39 | Screenshot 2021-07-08 at 17 36 04 |

The location is https://www.openstreetmap.org/#map=18/39.37976/-76.47098
Satelite view
Screenshot 2021-07-08 at 17 39 55

## What about tests
- gurka test is added
- I checked the diff with `pennsylvania.pbf`: the number of turnChannels increased by 2%(~4250 -> 4383), 95%+ percent are valid changes

## Rad tests
I rad tested the PR. The diff count is 42/2955(1.5%).  A few examples from diff:
### 1. Guidance
TurnChannel vs Ramp classification impacts guidance.
- we say "take the ramp/take the exit" for ramps
- we say "bear left/right" for turnChannels

Here you can see that `take exit` instruction is gone now
image
image

### 2.  Routing
TurnChannels raw speeds are increased during tile building(by 1.25x factor). So obviously it impacts routing https://github.com/valhalla/valhalla/blob/a7c3d4cf40eb9bdf6cdf2f2d745c0178d704eb91/src/mjolnir/speed_assigner.h#L16
(Note: this factor doesn't impact edgeCost of the turnChannel when historic speeds are added. To be experimented in the future...)
A few cases from rad tests
image
image
image

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/mjolnir/linkclassification.cc [hunk 1]
```diff
@@ -439,6 +423,311 @@ struct LinkGraphBuilder {
...
+    auto coords = json::array({json::fixed_t{p.lng(), 6}, json::fixed_t{p.lat(), 6}});
+    geometry->emplace("coordinates", std::move(coords));
+
+    feature->emplace("geometry", std::move(geometry));
+  }
+
+  return feature;
+}
+
+std::string VisualizeIntersection(
```
Comment (merkispavel) [inline src/mjolnir/linkclassification.cc:475]: helpful function to visualize slip lane result 
Screenshot 2021-07-08 at 18 03 25

File: src/mjolnir/linkclassification.cc [hunk 2]
```diff
@@ -439,6 +423,311 @@ struct LinkGraphBuilder {
...
+//  |   F
+//  | /
+//  A
+//  |
+// The idea is to check whether edges fit "triangle" pattern, i.e
+// when there another two ways from A to E:
+// - A-F-E though the slip lane
+// - A-B-C-D-E through the intersection C
+// e.g https://www.openstreetmap.org/way/5277672
+bool IsSlipLane(Data& data, SlipLaneInput input, double traverse_threshold) {
```
Comment (merkispavel) [inline src/mjolnir/linkclassification.cc:635]: The main function

File: src/mjolnir/linkclassification.cc [hunk 3]
```diff
@@ -439,6 +423,311 @@ struct LinkGraphBuilder {
...
+    return true;
+
+  // length is greater than threshold so we need further analysis
+  SlipLaneInput input = GetSlipLaneInput(data, link_edges);
+  if (input.Valid()) {
+#ifdef LOGGING_LEVEL_DEBUG
+    uint64_t way_id = (*data.ways[(*data.edges[link_edges[0]]).wayindex_]).way_id();
+    LOG_DEBUG("Link edges with way_id=" + std::to_string(way_id));
+#endif
+    return IsSlipLane(data, input, 2 * total_length);
```
Comment (genadz) [inline src/mjolnir/linkclassification.cc:726]: pls, add comment here why our length threshold equals to `2 * total_length`

File: test/gurka/test_ramps_tc.cc [hunk 4]
```diff
@@ -476,4 +476,37 @@ TEST(LinkReclassification, test_acyclic_graph) {
...
+           W____C____D___E____N
+                |        /
+                |      ^
+                B     /
+                |   F
+                | /
+                A
+                |
+                X
+    )";
```
Comment (genadz) [inline test/gurka/test_ramps_tc.cc:497]: could you also add a test example for left slip lane ?

File: src/mjolnir/linkclassification.cc [hunk 5]
```diff
@@ -439,6 +423,311 @@ struct LinkGraphBuilder {
...
+      if (delta < min_angle_delta) {
+        min_angle_delta = delta;
+        best_edge_idx = to.second;
+      }
+    }
+    return best_edge_idx;
+  };
+
+  SlipLaneInput res;
+  Edge first_link_edge = *data.edges[link_edges.back()];
```
Comment (genadz) [inline src/mjolnir/linkclassification.cc:692]: it's not clear why **first** link edge is the last one from `link_edges` list. We should add a comment about that, or reverse `link_edges` list before call `GetSlipLaneInput` function

**Review by genadz — APPROVED** [review]

**Review by dgearhart — APPROVED** [review]
thanks @merkispavel  - the RAD results looked good

**Pull request merged**